src/main/java/com/google/devtools/build/lib/packages/TargetUtils.java

```diff
@@ -339,14 +346,38 @@ private static boolean isExplicitDependency(Rule rule, Label label) {
   }
 
   /**
-   * Returns a predicate to be used for test tag filtering, i.e., that only accepts tests that match
-   * all of the required tags and none of the excluded tags.
+   * Returns a predicate to be used for test tag filtering based on either a list of tags or a Starlark expression.
+   * If a list of tags is provided the predicates evaluate to true when the test tags match all of the required tags and
+   * none of the excluded tags.
+   * A Starlark expression should be provided as a single string in the parameter tagFilterList and is restricted to the
+   * operators specified in evalExpression.
+   * If a Starlark expression has invalid syntax or uses restricted operators a predicate that always evaluates to true
+   * is returned.
    */
   public static Predicate<Target> tagFilter(List<String> tagFilterList) {
     Pair<Collection<String>, Collection<String>> tagLists =
         TestTargetUtils.sortTagsBySense(tagFilterList);
     final Collection<String> requiredTags = tagLists.first;
     final Collection<String> excludedTags = tagLists.second;
+
+    // If there is only one required tag and none excluded tags we can be sure that it is safe to handle this as a
+    // Starlark expression.
+    // If the comparison fails we will handle it as a list of tags.
+    if (requiredTags.size() == 1 && excludedTags.size() == 0) {
```

Comment:
It would be much clearer to have the caller provide a string (not a list of strings already split on commas) and parse it here. Then this condition becomes simply !str.contains(',').  If the string does contain a comma, it's trivial to split it and return an evaluator for an array of implicitly conjoined tags, or emit an error message "invalid boolean formula: a,b,c (did you mean 'a and b and c'?)".